File: english/data_processing/lessons/code/pc2grid.m
#! /usr/bin/octave -qf
format long;
% process command line arguments
if (nargin > 0)
  arg_list = argv();
  pcfile = arg_list{1};       % point cloud
  if (nargin > 1)
    dx = str2num(arg_list{2});  % grid step
  else
    dx = 1;                   % default cell size
  end
  if (nargin > 2)
    fname = arg_list{3};        % function to use
  else
    fname = "min";
  end
  if (strncmp(fname, "min", 2))
    fu = @min;
  elseif (strncmp(fname, "max", 2))
    fu = @max;
  elseif (strncmp(fname, "mean", 3))
    fu = @mean;
  elseif (strncmp(fname, "median", 3))
    fu = @median;
  elseif (strncmp(fname, "numel", 2))
    fu = @numel;
  end
else
  printf("Usage pc2grid.m point_cloud_file [resolution] [min/max/mean/median/num]\n");
  exit(1)
end
no_data = -9999.0;
start_time = time();
% load point cloud and remove extra columns
points = load(pcfile)(:,1:3);
printf("--- reading %.2f seconds ---\n", time() - start_time);
start_time1 = time();
d = [dx, dx, dx];
minp = min(points, [], 1);    % get min coords
maxp = max(points, [], 1);    % get max coords
minp = floor(minp ./ d) .* d;
maxp = ceil(maxp ./d) .* d;
n = uint32((maxp .- minp) ./ d);% grid sizes
% calculate row and column index to points
indexes = uint32(floor((points .- minp) ./ d)) .+ 1;
indexes = [indexes (1:1:rows(points))'];
printf("--- indexing %.2f seconds ---\n", time() - start_time1);
% create grid selecting point in cells
% headers for grid file
start_time2 = time();
f = fopen(strcat(pcfile, ".asc"), "w");
fprintf(f, "ncols %d\n", n(1));
fprintf(f, "nrows %d\n", n(2));
fprintf(f, "xllcorner %.3f\n", minp(1));
fprintf(f, "yllcorner %.3f\n", minp(2));
fprintf(f, "cellsize %.3f\n", d(1));
fprintf(f, "nodata_value %.3f\n", no_data);
for i = n(2):-1:1             % rows from top to down
  idx = (indexes(:, 2) == i); % select points in ith row
  ii = indexes(idx, :);
  ppi = points(idx, :);
  for j = 1:n(1)
   idy = (ii(:, 1) == j);    % select points in jth cell
    ppj = ppi(idy, :);
    if (rows(ppj))
      gr = fu(ppj(:, 3));
      fprintf(f, "%.3f ", gr);
    else
      fprintf(f, "%.3f ", no_data);      
    end
  end
  fprintf(f, "\n");
end
fclose(f);
printf("--- griding1 %.2f seconds ---\n", time() - start_time2);
% create grid sorting points by bucket indices
sorted_indexes = sortrows(indexes, [-2, 1]);
start_time3 = time();
% headers for grid file
f = fopen(strcat(pcfile, "_1.asc"), "w");
fprintf(f, "ncols %d\n", n(1));
fprintf(f, "nrows %d\n", n(2));
fprintf(f, "xllcorner %.3f\n", minp(1));
fprintf(f, "yllcorner %.3f\n", minp(2));
fprintf(f, "cellsize %.3f\n", d(1));
fprintf(f, "nodata_value %.3f\n", no_data);
% buffer for a row of grid
grid = ones(n(1), 1) * no_data;
i = sorted_indexes(1,2);
j = sorted_indexes(1,1);
start = 1;
m = rows(sorted_indexes);
for k = 1:m
  % grid distance in row order of cells
  gd = (i - sorted_indexes(k, 2)) * n(1) + sorted_indexes(k, 1) - j;
  if gd
    grid(j) = fu(points(sorted_indexes(start:k-1, 4), 3));
    for ii = sorted_indexes(k, 2):i-1
      for jj = 1:n(1)
          fprintf(f, "%.3f ", grid(jj));
      end
      fprintf(f, "\n");  
      grid = ones(n(1), 1) * no_data;    
    end
    j = sorted_indexes(k,1);
    i = sorted_indexes(k, 2);
    start = k;
  end
end
% set last bucket
grid(j) = fu(points(sorted_indexes(start:m, 4), 3));
for jj = 1:n(1)
  fprintf(f, "%.3f ", grid(jj));
end
fprintf(f, "\n");  
fclose(f);
printf("--- griding2 %.2f seconds ---\n", time() - start_time3);
% collect points in the same cell
start_time4 = time();
% headers for grid file
f = fopen(strcat(pcfile, "_2.asc"), "w");
fprintf(f, "ncols %d\n", n(1));
fprintf(f, "nrows %d\n", n(2));
fprintf(f, "xllcorner %.3f\n", minp(1));
fprintf(f, "yllcorner %.3f\n", minp(2));
fprintf(f, "cellsize %.3f\n", d(1));
fprintf(f, "nodata_value %.3f\n", no_data);
grid = cell(n(2),n(1));
for k = 1:m
  grid{indexes(k, 2), indexes(k, 1)} = [grid{indexes(k,2), indexes(k, 1)}, points(k, 3)];
end
for i = n(2):-1:1             % rows from top to down
  for j = 1:n(1)
    if columns(grid{i, j})
      fprintf(f, "%.3f ", fu(grid{i, j}));
    else
      fprintf(f, "%.3f ", no_data);
    end
  end
  fprintf(f, "\n");
end
printf("--- griding3 %.2f seconds ---\n", time() - start_time4);

File: english/data_processing/lessons/code/pc2grid.py
#! /usr/bin/env python3
"""
   Sample code to demonstrate effectiveness of algorithms
"""
import time
import sys
import os
import numpy as np

def num(an_array):
    """ number of rows in a numpy array """
    return an_array.size

ARGC = len(sys.argv)
if ARGC == 1:
    print("usage: {} point_file [step] [min/max/mean/median/num]\n".format(sys.argv[0]))
    sys.exit(1)
fname = sys.argv[1]
# grid sizes
DX = 0.1
if ARGC > 2:
    DX = float(sys.argv[2])
d = np.array([DX, DX, DX])
# function for interpolation
fu = np.amin
if ARGC > 3:
    if sys.argv[3][0:2].lower() == "mi":        # minimum Z
        fu = np.amin
    elif sys.argv[3][0:2].lower() == "ma":      # maximum Z
        fu = np.amax
    elif sys.argv[3][0:3].lower() == "mea":     # mean Z
        fu = np.mean
    elif sys.argv[3][0:3].lower() == "med":     # median Z
        fu = np.median
    elif sys.argv[3][0:2].lower() == "nu":      # number of points
        fu = num
no_data = -9999.0       # nodata in generated ascii grid

# load ascii point cloud x y z ...
start_time = time.time()
points = np.loadtxt(fname)
print("--- reading {} seconds ---".format((time.time() - start_time)))

start_time1 = time.time()
# delete extra columns
points = np.delete(points, np.s_[3:], axis=1)
# get extent
minp = np.amin(points, axis=0)
maxp = np.amax(points, axis=0)
# round to grid values
minp = np.floor(minp / d) * d
maxp = np.ceil(maxp / d) * d
# number of buckets along the axes
n = ((maxp - minp) / d).astype('int32')
# creating index bucket rows and columns
indexes = ((points - minp) / d).astype('int32')
# add point indices in points array
indexes = np.append(indexes, np.arange(points.shape[0], dtype='int32').reshape(points.shape[0], 1), axis=1)
print("--- indexing {} seconds ---".format((time.time() - start_time1)))

# create grid selecting points in grid cells
# create ESRI ASCII grid output
start_time2 = time.time()
oname = os.path.splitext(fname)[0] + ".asc"
f = open(oname, "w")
f.write("ncols {}\n".format(n[0]))
f.write("nrows {}\n".format(n[1]))
f.write("xllcorner {:.3f}\n".format(minp[0]))
f.write("yllcorner {:.3f}\n".format(minp[1]))
f.write("cellsize {:.3f}\n".format(d[0]))
f.write("nodata_value {:.3f}\n".format(no_data))

for i in reversed(range(n[1])):
    ii = indexes[indexes[:, 1] == i]     # points in the ith row of buckets
    for j in range(n[0]):
        jj = ii[ii[:, 0] == j]           # points in the single bucket
        pp = points[jj[:, 3]]
        if pp.size:
            gr = fu(pp[:, 2])     # apply min/max/mean/median
            f.write("{:.3f} ".format(gr))
        else:
            f.write("{:.3f} ".format(no_data))
    f.write("\n")
f.close()
print("--- griding1 {} seconds ---".format((time.time() - start_time2)))
# scanning sorted points
start_time3 = time.time()
# sorting indices array by bucket indices, decreasing rows and increasing cols
sorted_indexes = indexes[np.lexsort((indexes[:, 3], indexes[:, 2],
                                     indexes[:, 0], -indexes[:, 1]))]
oname = os.path.splitext(fname)[0] + "_1.asc"
f = open(oname, "w")
f.write("ncols {}\n".format(n[0]))
f.write("nrows {}\n".format(n[1]))
f.write("xllcorner {:.3f}\n".format(minp[0]))
f.write("yllcorner {:.3f}\n".format(minp[1]))
f.write("cellsize {:.3f}\n".format(d[0]))
f.write("nodata_value {:.3f}\n".format(no_data))
# buffer for a row of grid
grid = np.empty((n[0]))
grid.fill(no_data)
i = sorted_indexes[0, 1]
j = sorted_indexes[0, 0]
start = 0
m = sorted_indexes.shape[0]        # number of points
for k in range(m):
    # grid distance in row order of cells
    gd = -sorted_indexes[k, 1] * n[0] + sorted_indexes[k, 0] + i * n[0] - j
    if gd:
        # new bucket reached
        try:                    # TODO index out of range error
            grid[j] = fu(points[sorted_indexes[start:k, 3], 2])
        except IndexError:
            pass
        for ii in range(sorted_indexes[k, 1], i):
            for jj in range(n[0]):
                f.write("{:.3f} ".format(grid[jj]))
            f.write("\n")
            grid.fill(no_data)      # initialize row buffer
        j = sorted_indexes[k, 0]
        i = sorted_indexes[k, 1]
        start = k
# set last bucket
try:
    grid[j] = fu(points[sorted_indexes[start:m, 3], 2])
except IndexError:
    pass
for jj in range(n[0]):              # output a row to grid
    f.write("{:.3f} ".format(grid[jj]))
f.write("\n")
f.close()

print("--- griding2 {} seconds ---".format((time.time() - start_time3)))
start_time4 = time.time()
# scanning unsorted points
grid = {}
for i in range(n[1]):
    for j in range(n[0]):
        grid[(i, j)] = []    # initialize dict with empty lists
m = indexes.shape[0]        # number of points
for k in range(m):
    try:                     # TODO index out of range error
        grid[(indexes[k, 1], indexes[k, 0])].append(points[k, 2])
    except KeyError:
        pass

oname = os.path.splitext(fname)[0] + "_2.asc"
# output
f = open(oname, "w")
f.write("ncols {}\n".format(n[0]))
f.write("nrows {}\n".format(n[1]))
f.write("xllcorner {:.3f}\n".format(minp[0]))
f.write("yllcorner {:.3f}\n".format(minp[1]))
f.write("cellsize {:.3f}\n".format(d[0]))
f.write("nodata_value {:.3f}\n".format(no_data))
for i in reversed(range(n[1])):
    for j in range(n[0]):
        if len(grid[i, j]) > 0:
            f.write("{:.3f} ".format(fu(np.array(grid[(i, j)]))))
        else:
            f.write("{:.3f} ".format(no_data))
    f.write("\n")
f.close()
print("--- griding3 {} seconds ---".format((time.time() - start_time4)))
print("--- total {} seconds ---".format((time.time() - start_time)))
